Add documentation on testing Chrome Extensions (#7479)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/tut_puppeteer-testing/index.md

---
layout: "layouts/doc-post.njk"
title: "Tutorial: Testing Chrome Extensions with Puppeteer"
seoTitle: "Tutorial: Testing Chrome Extensions with Puppeteer"
date: 2023-10-09
description: How to write an automated test for Chrome Etensions in Puppeteer.
---

_For general advice on testing extensions, see
[Automated testing for Chrome Extensions][automated-testing]._

> [Puppeteer][puppeteer] provides a framework for building automated tests of websites, which also

Reviewer: Add a sentence explaining what this tutorial teaches.

includes the ability to test Chrome Extensions. These are high-level integration tests that test the
functionality of a website or extension once it has been built into the final product.

## Before you start

Clone or download the [chrome-extensions-samples][samples-repo] repository. We'll use the history
API demo in `api-samples/history/showHistory` as our test extension.

You'll also need to install [Node.JS][node] which is the runtime Puppeteer is built on.

> ## Writing your test

Reviewer: Did you consider using Jest here?

Author: As a test runner or for unit tests? I think a separate unit testing tutorial is something we could look at in the future, although that was covered at a high level on the other page. I don't think we want to show it as a test runner because this tutorial would quickly explode in complexity as you start needing to explain parallelism, how to setup and tear things down. At least for this first PR I wanted to focus on the bits that are specific to extensions and where good tutorials don't already exist. You could hopefully use this tutorial to set things up in any test runner of your choice.

Reviewer: Just as a test runner. Setup / tear down + parallelism are important things we should cover here. If it adds too much overhead I'm fine with starting with this approach.

Author: Updated the tutorial with Jest :)

### Step 1: Start your Node.JS project

We need to set up a basic Node.JS project. In a new folder, create a `package.json` file with the
following:

```js
{
  "name": "puppeteer-demo",
  "version": "1.0"
}
```

Similar to an extension's manifest.json file, this file is required by all Node projects.

### Step 2: Install Puppeteer

> Run `npm install puppeteer` to add Puppeteer as a dependency. It will be automatically added to your

Reviewer: ```suggestion
Run `npm install puppeteer --save-dev` to add Puppeteer as a dependency. It will be automatically added to your
```

Otherwise dependency won't be added

Author: I believe NPM now saves to the main `dependencies` key by default: https://docs.npmjs.com/cli/v8/commands/npm-install#save

I think `dependencies` actually makes sense over `devDependencies` here because Playwright is required to run the tests, not just to build or develop them. That's how I usually think about it anyway (there's admittedly some grey area for local tooling and either works in practice).

Reviewer: > I believe NPM now saves to the main dependencies key by default: https://docs.npmjs.com/cli/v8/commands/npm-install#save

TIL

> I think dependencies actually makes sense over devDependencies here because Playwright is required to run the tests, not just to build or develop them. That's how I usually think about it anyway (there's admittedly some grey area for local tooling and either works in practice).

Usually, deps required for running tests is definitely a devDependency (my rule of thumb is always to think about is this going to be required in prod or not). In this particular though, as it is a standalone testing project, I agree that it makes more sense as a normal dependency.

`package.json` file.

### Step 3: Create an entry point

In a new file called `index.js`, add the following:

```js
const puppeteer = require('puppeteer');

const EXTENSION_PATH = '../../api-samples/history/showHistory';
const EXTENSION_ID = 'jkomgjfbbjocikdmilgaehbfpllalmia';

async function run() {
  const browser = await puppeteer.launch({

>     headless: false,

Reviewer: Why not `true`?

    args: [
      `--disable-extensions-except=${EXTENSION_PATH}`,
      `--load-extension=${EXTENSION_PATH}`
    ]
  });
}

run();
```

Run `node index.js` and you should see Chrome open with your extension loaded.

### Step 4: Add an alias

To make running the tests easier, add an alias to your `package.json` file:

```js
{
  "name": "puppeteer-demo",
  "version": "1.0",
  "dependencies": {
    "puppeteer": "^21.3.6"
  },
  "scripts": {

>     "start": "node index.js"

Reviewer: Why not `test`?

Author: Just personal preference here, I can change it if you feel strongly. Since this whole NPM project **is** the test I think having `start` run the tests makes sense. I usually think of `test` as testing the NPM project, which would in this case mean testing the tests.

Reviewer: No strong preferences.  It makes it easier to always use the same task name because it makes test execution easier in a multi-project setup.

---
layout: "layouts/doc-post.njk"
title: "Tutorial: Testing Chrome Extensions with Puppeteer"
seoTitle: "Tutorial: Testing Chrome Extensions with Puppeteer"
date: 2023-10-09
description: How to write an automated test for Chrome Etensions in Puppeteer.
---

_For general advice on testing extensions, see

> [Automated testing for Chrome Extensions][automated-testing]._

Reviewer 2: This feels like it should go after the intro to the tutorial and maybe an aside? Your call

Author: I think this is best so will leave for now, but open to changing it in the future.

[Puppeteer][puppeteer] provides a framework for building automated tests of websites, which also
includes the ability to test Chrome Extensions. These are high-level end-to-end tests that test the
functionality of a website or extension once it has been built into the final product. In this
tutorial, we demonstrate how to write a basic test for an extension from our samples repository.

> ## Before you start

Reviewer 2: ```suggestion
## Before you start {: #prereq }
```

Clone or download the [chrome-extensions-samples][samples-repo] repository. We'll use the history
API demo in `api-samples/history/showHistory` as our test extension.

You'll also need to install [Node.JS][node] which is the runtime Puppeteer is built on.

> ## Writing your test

Reviewer 2: ```suggestion
## Writing your test {: #write-test }
```

> ### Step 1: Start your Node.JS project

Reviewer 2: ```suggestion
### Step 1: Start your Node.JS project {: #step-1 }
```

We need to set up a basic Node.JS project. In a new folder, create a `package.json` file with the
following:

```js
{
  "name": "puppeteer-demo",
  "version": "1.0"
}

> ```

Reviewer: This is a bit quicker for setting up a project:

```
npm init -y
```

> Similar to an extension's manifest.json file, this file is required by all Node projects.

Reviewer 2: ```suggestion
Similar to an extension's `manifest.json` file, this file is required by all Node projects.
```

> ### Step 2: Install Puppeteer and Jest

Reviewer 2: ```suggestion
### Step 2: Install Puppeteer and Jest {: #step-2 }

```

Run `npm install puppeteer jest` to add Puppeteer and Jest as dependencies. They will be
automatically added to your `package.json` file.

It is possible to write standalone Puppeteer tests, but we'll use Jest as a test runner to provide
some additional structure to our code.

> ### Step 3: Create an entry point

Reviewer 2: ```suggestion
### Step 3: Create an entry point {: #step-3 }

```

> In a new file called `index.test.js`, add the following:

Reviewer 2: ```suggestion
Create a new file called `index.test.js` and add the following code:
```

{% Aside %}
Jest adds `beforeEach` and `afterEach` to our environment when running the tests, so we do not need
to import them. However, you may need to configure [ESLint][eslint] or [TypeScript][typescript]
accordingly.
{% endAside %}

```js
const puppeteer = require('puppeteer');

const EXTENSION_PATH = '../../api-samples/history/showHistory';
const EXTENSION_ID = 'jkomgjfbbjocikdmilgaehbfpllalmia';

let browser;

beforeEach(async () => {
  // TODO: Launch the browser.
});

afterEach(async () => {
  // TODO: Close the browser.
});
```

> ### Step 4: Launch the browser

Reviewer 2: ```suggestion
### Step 4: Launch the browser {: #step-4 }

```

{% Aside %}
In this example, `headless` is set to `false`. This causes the browser window to be visible while
the tests are running which can be helpful during development. Outside of development, consider
setting it to `'new'` which uses Chrome's [new headless mode][new-headless].
{% endAside %}

Update `beforeEach` and `afterEach` to launch and close the browser. When running many tests, you
may wish to consider using the same browser. However, this is generally discouraged as it reduces
the isolation between your tests and may cause one test to impact the outcome of another.

```js
beforeEach(async () => {
  browser = await puppeteer.launch({
    headless: false,
    args: [
      `--disable-extensions-except=${EXTENSION_PATH}`,
      `--load-extension=${EXTENSION_PATH}`
    ]
  });
});

afterEach(async () => {
  await browser.close();
  browser = undefined;
});
```

> ### Step 5: Add an alias

Reviewer 2: ```suggestion
### Step 5: Add an alias {: #step-5 }

```

To make running the tests easier, add an alias to your `package.json` file:

```js
{
  "name": "puppeteer-demo",
  "version": "1.0",
  "dependencies": {
    "puppeteer": "^21.3.6"
  },
  "scripts": {
    "start": "jest ."
  }
}
```

This will run all files ending in `.test.js` in the current directory.

> ### Step 6: Open the popup

Reviewer 2: ```suggestion
### Step 6: Open the popup {: #step-6 }

```

Let's add a basic test that opens the popup in a new page. We need to do this because Puppeteer
does not support accessing an extension popup from the popup window. Add the following code:

> 

Reviewer 2: Add file label

```js
test('popup renders correctly', async () => {
  const page = await browser.newPage();
  await page.goto(`chrome-extension://${EXTENSION_ID}/popup.html`);
});
```

> The method `browser.newPage()` actually supports an optional URL parameter, but unfortunately this

Reviewer 2: Optional: add a link to `browser.newPage()`?

Author: Realised this paragraph was wrong so dropped it entirely.

does not support URLs with the `chrome-extension://` scheme. Consequently, we first open a page and
then call `page.goto()` to navigate.

> ### Step 7: Assert the current state

Reviewer 2: ```suggestion
### Step 7: Assert the current state {: #step-7 }
```

Let's assert something, so that our test can fail if the extension isn't behaving as expected. We
know that our popup should show recently visited pages, so let's check that we see one:

```js
test('popup renders correctly', async () => {
  const page = await browser.newPage();
  await page.goto(`chrome-extension://${EXTENSION_ID}/popup.html`);

  const list = await page.$('ul');
  const children = await list.$$('li');

  expect(children.length).toBe(1);
});
```

> ### Step 8: Run your test

Reviewer 2: ```suggestion
### Step 8: Run your test {: #step-8 }
```

To run the test, use `npm start`. You should see output indicating that your test passed.

You can see the [full project][full-project] in our chrome-extensions-samples repository.

> ## Next Steps

Reviewer 2: ```suggestion
## Next Steps {: #next-steps }
```

---
layout: "layouts/doc-post.njk"
title: "Tutorial: Testing Chrome Extensions with Puppeteer"
seoTitle: "Tutorial: Testing Chrome Extensions with Puppeteer"
date: 2023-10-09
description: How to write an automated test for Chrome Etensions in Puppeteer.
---

> _For general advice on testing extensions, see
> [Automated testing for Chrome Extensions][automated-testing]._

Reviewer 2: ```suggestion
_For general advice on testing extensions, see
[Automated testing for Chrome Extensions][automated-testing] and [Unit testing in Chrome Extensions][unit-testing]._
```

  const list = await page.$('ul');
  const children = await list.$$('li');

  expect(children.length).toBe(1);
});
```

### Step 8: Run your test {: #step-8 }

To run the test, use `npm start`. You should see output indicating that your test passed.

You can see the [full project][full-project] in our chrome-extensions-samples repository.

## Next Steps {: #next-steps }

After mastering the basics, try building a test suite for your own extension. The Puppeteer
[API reference][api-reference] contains more information about what's possible - there are many
capabilities not described here.

[automated-testing]: /docs/extensions/mv3/automated-testing
[puppeteer]: https://github.com/puppeteer/puppeteer
[samples-repo]: https://github.com/GoogleChrome/chrome-extensions-samples
[node]: https://nodejs.org/
[new-headless]: https://developer.chrome.com/articles/new-headless/
[full-project]: https://github.com/GoogleChrome/chrome-extensions-samples/tree/main/functional-samples/tutorial.puppeteer
[api-reference]: https://pptr.dev/api
[consistent-id]: /docs/extensions/mv3/automated-testing/#setting-an-extension-id
[eslint]: https://www.npmjs.com/package/eslint-plugin-jest

> [typescript]: https://jestjs.io/docs/getting-started#type-definitions

Reviewer 2: ```suggestion
[typescript]: https://jestjs.io/docs/getting-started#type-definitions
[unit-testing]: /docs/extensions/mv3/unit-testing

```

---
layout: "layouts/doc-post.njk"
title: "Tutorial: Testing Chrome Extensions with Puppeteer"
seoTitle: "Tutorial: Testing Chrome Extensions with Puppeteer"

> date: 2023-10-09

Reviewer 2: ```suggestion
date: 2023-10-12
```

> description: How to write an automated test for Chrome Etensions in Puppeteer.

Reviewer 2: ```suggestion
description: How to write an automated test for Chrome Extensions using Puppeteer.
```

---

_For general advice on testing extensions, see

> [Automated testing for Chrome Extensions][automated-testing] and [Unit testing in Chrome Extensions][unit-testing]._

Reviewer 2: ```suggestion
[End-to-end testing of Chrome Extensions][automated-testing] and [Unit testing in Chrome Extensions][unit-testing]._
```

test('popup renders correctly', async () => {
  const page = await browser.newPage();
  await page.goto(`chrome-extension://${EXTENSION_ID}/popup.html`);
});
```

### Step 7: Assert the current state {: #step-7 }

Let's assert something, so that our test can fail if the extension isn't behaving as expected. We
know that our popup should show recently visited pages, so let's check that we see one:

{% Label %}index.test.js:{% endLabel %}

```js
test('popup renders correctly', async () => {
  const page = await browser.newPage();
  await page.goto(`chrome-extension://${EXTENSION_ID}/popup.html`);

  const list = await page.$('ul');
  const children = await list.$$('li');

  expect(children.length).toBe(1);
});
```

### Step 8: Run your test {: #step-8 }

To run the test, use `npm start`. You should see output indicating that your test passed.

You can see the [full project][full-project] in our chrome-extensions-samples repository.

## Next Steps {: #next-steps }

After mastering the basics, try building a test suite for your own extension. The Puppeteer
[API reference][api-reference] contains more information about what's possible - there are many
capabilities not described here.

> [automated-testing]: /docs/extensions/mv3/automated-testing

Reviewer 2: ```suggestion
[automated-testing]: /docs/extensions/mv3/end-to-end-testing/
```
